spaceroombas/data/state.py

```diff
@@ -259,20 +266,59 @@ def __init__(self):
         self.NEW_ROBOT_FLAG = False
 
     def add_player(self, player_id):
-        sector = self.map.get_sector(
-            "0,0")  # TODO implement logic to spawn new player in their own starting sector/for now to base
-        playerState = PlayerState(player_id, sector, self.parser, self.transpiler)
+        x, y = self.choose_spawn_sector()
+        sector_id = MapSector.form_sector_id(x, y)
+
+        self.map.generate_map_sector(sector_id)
+        sector = self.map.get_sector(sector_id)
 
-        if player_id in self.players:  ## TODO implement logic here for orphan clients
+        playerState = PlayerState(
+            player_id, sector, self.parser, self.transpiler)
+
+        if player_id in self.players:  # TODO implement logic here for orphan clients
             raise PlayerExistsError("Player already exists")
 
         self.players[player_id] = playerState
-        x = 20  # TODO implement logic to spawn starting robot not in a used tile
+
+        x = 20
         y = 20
+        while False == self.map.check_tile_available(EntityLocation(sector, x, y)):
+            logging.warning("tile in use", x, y, sector_id)
+            x += 1
+
         self.add_robot(player_id, EntityLocation(sector, x, y))
         logging.info("Player added: %s" % (player_id))
 
-    def add_robot(self, player_id, location: EntityLocation):  # TODO set up so the robot is spawned in valid location
+    def choose_spawn_sector(self):
+        # returns false if any point is with in the distance
+        def check_distance(x1, y1, previous, dist):
+            for p in previous:
+                if dist > abs(x1-p[0])+abs(y1-p[1]):  # manhattan distance
+                    return False
+            return True
+
+        indexs = []
+        sectors = self.map.get_sector_ids()
+
+        for s in sectors:
+            indexs.append(MapSector.parse_id(s))
+
+        xdelta = 0
+        ydelta = 0
+        x = 0
+        y = 0
+        distance = random.randint(3, 6)
+        while xdelta == 0 and ydelta == 0:
+            xdelta = random.randint(-1, 1)
+            ydelta = random.randint(-1, 1)
+
+        while check_distance(x, y, indexs, distance) == False:
+            x += xdelta
+            y += ydelta
+        return x, y
```

Comment:
should this return an EntityLocation?

Reply:
it could, sector isnt relevant in this context though as they should all spawn in the same sector, not go else where